pinot-core/src/main/java/org/apache/pinot/core/query/reduce/GapFillGroupByDataTableReducer.java

```diff
@@ -0,0 +1,483 @@
+        Object[] gapfillRow = new Object[numResultColumns];
+        if (resultColumnDataTypes[0] == ColumnDataType.LONG) {
+          gapfillRow[0] = Long.valueOf(_dateTimeFormatter.fromMillisToFormat(time));
+        } else {
+          gapfillRow[0] = _dateTimeFormatter.fromMillisToFormat(time);
+        }
+        System.arraycopy(key.getValues(), 0, gapfillRow, 1, _numOfKeyColumns);
+
+        for (int i = _numOfKeyColumns + 1; i < numResultColumns; i++) {
+          gapfillRow[i] = getFillValue(i, key, resultColumnDataTypes[i]);
+        }
+        gapfillResultRows.add(gapfillRow);
+        if (gapfillResultRows.size() == limit) {
+          return gapfillResultRows;
+        }
+      }
+    }
+    return gapfillResultRows;
+  }
+
+  Object getFillValue(int columIndex, Object key, ColumnDataType dataType) {
+    ExpressionContext expressionContext = _queryContext.getSelectExpressions().get(columIndex);
+    if (expressionContext.getFunction() != null
+        && expressionContext.getFunction().getFunctionName().equalsIgnoreCase("fill")) {
+      List<ExpressionContext> args = expressionContext.getFunction().getArguments();
+      if (args.get(1).getLiteral() == null) {
+        throw new UnsupportedOperationException("Wrong Sql.");
+      }
+      FillType fillType = FillType.valueOf(args.get(1).getLiteral());
+      if (fillType == FillType.FILL_DEFAULT_VALUE) {
+        // TODO: may fill the default value from sql in the future.
+        return getDefaultValue(dataType);
+      } else if (fillType == FillType.FILL_PREVIOUS_VALUE) {
+        Object[] row = _previous.get(key);
+        if (row != null) {
+          return row[columIndex];
+        } else {
+          return getDefaultValue(dataType);
+        }
+      } else {
+        throw new UnsupportedOperationException("unsupported fill type.");
+      }
+    } else {
+      return getDefaultValue(dataType);
+    }
+  }
+
+  enum FillType {
+    FILL_DEFAULT_VALUE,
+    FILL_PREVIOUS_VALUE,
+  }
+
+  /**
+   * The default value for each column type.
+   */
+  private Serializable getDefaultValue(ColumnDataType dataType) {
+    switch (dataType) {
+      // Single-value column
+      case INT:
+      case LONG:
+      case FLOAT:
+      case DOUBLE:
+      case BOOLEAN:
+      case TIMESTAMP:
+        return dataType.convertAndFormat(0);
+      case STRING:
+      case JSON:
+      case BYTES:
+        return "";
+      case INT_ARRAY:
+        return new int[0];
+      case LONG_ARRAY:
+        return new long[0];
+      case FLOAT_ARRAY:
+        return new float[0];
+      case DOUBLE_ARRAY:
+        return new double[0];
+      case STRING_ARRAY:
+      case TIMESTAMP_ARRAY:
+        return new String[0];
+      case BOOLEAN_ARRAY:
+        return new boolean[0];
+      case BYTES_ARRAY:
+        return new byte[0][0];
+      default:
+        throw new IllegalStateException(String.format("Cannot provide the default value for the type: %s", dataType));
+    }
+  }
```

Comment:
Suggest moving these 2 methods into the utils class since they can be shared for pre-aggregate gapfill

Reply:
Fixed.